// react-native-otp-inputs/lib/module/index.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import Clipboard from '@react-native-community/clipboard';
import React, { forwardRef, useCallback, useEffect, useImperativeHandle, useReducer, useRef } from 'react';
import { Keyboard, Platform, StyleSheet, View } from 'react-native';
import OtpInput from './OtpInput';
import { OtpInputsRef } from './types';
import { fillOtpCode } from './helpers';
const ACTION_TYPES = {
  setHandleChange: 'setHandleChange',
  setOtpTextForIndex: 'setOtpTextForIndex',
  setOtpCode: 'setOtpCode',
  clearOtp: 'clearOtp',
  setHasKeySupport: 'setHasKeySupport'
};

const reducer = (state, {
  type,
  payload
}) => {
  switch (type) {
    case ACTION_TYPES.setOtpTextForIndex:
      {
        const otpCode = _objectSpread(_objectSpread({}, state.otpCode), {}, {
          ["".concat(payload.index)]: payload.text
        });

        state.handleChange(Object.values(otpCode).join(''));
        return _objectSpread(_objectSpread({}, state), {}, {
          otpCode
        });
      }

    case ACTION_TYPES.setOtpCode:
      {
        const otpCode = fillOtpCode(payload.numberOfInputs, payload.code);
        state.handleChange(Object.values(otpCode).join(''));
        return _objectSpread(_objectSpread({}, state), {}, {
          otpCode
        });
      }

    case ACTION_TYPES.clearOtp:
      {
        const otpCode = fillOtpCode(payload);
        state.handleChange(Object.values(otpCode).join(''));
        return _objectSpread(_objectSpread({}, state), {}, {
          otpCode
        });
      }

    case ACTION_TYPES.setHandleChange:
      {
        return _objectSpread(_objectSpread({}, state), {}, {
          handleChange: payload
        });
      }

    case ACTION_TYPES.setHasKeySupport:
      {
        return _objectSpread(_objectSpread({}, state), {}, {
          hasKeySupport: payload
        });
      }

    default:
      return state;
  }
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between'
  }
});
const OtpInputs = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    autofillFromClipboard = true,
    autoCapitalize = 'none',
    clearTextOnFocus = false,
    defaultValue,
    focusStyles,
    handleChange = console.log,
    inputContainerStyles,
    inputStyles,
    isRTL = false,
    keyboardType = 'phone-pad',
    numberOfInputs = 4,
    placeholder = '',
    secureTextEntry = false,
    selectTextOnFocus = true,
    style,
    testIDPrefix = 'otpInput'
  } = _ref,
      restProps = _objectWithoutProperties(_ref, ["autofillFromClipboard", "autoCapitalize", "clearTextOnFocus", "defaultValue", "focusStyles", "handleChange", "inputContainerStyles", "inputStyles", "isRTL", "keyboardType", "numberOfInputs", "placeholder", "secureTextEntry", "selectTextOnFocus", "style", "testIDPrefix"]);

  const previousCopiedText = useRef('');
  const inputs = useRef([]);
  const [{
    otpCode,
    hasKeySupport
  }, dispatch] = useReducer(reducer, {}, () => ({
    otpCode: fillOtpCode(numberOfInputs, defaultValue),
    handleChange,
    hasKeySupport: Platform.OS === 'ios'
  }));
  useEffect(() => {
    dispatch({
      type: ACTION_TYPES.setHandleChange,
      payload: handleChange
    });
  }, [handleChange]);
  useImperativeHandle(ref, () => ({
    reset: () => {
      dispatch({
        type: ACTION_TYPES.clearOtp,
        payload: numberOfInputs
      });
      inputs.current.forEach(input => {
        var _input$current;

        return input === null || input === void 0 ? void 0 : (_input$current = input.current) === null || _input$current === void 0 ? void 0 : _input$current.clear();
      });
      previousCopiedText.current = '';
      Clipboard.setString('');
    },
    focus: () => {
      var _firstInput$current;

      const firstInput = inputs.current[0];
      firstInput === null || firstInput === void 0 ? void 0 : (_firstInput$current = firstInput.current) === null || _firstInput$current === void 0 ? void 0 : _firstInput$current.focus();
    }
  }), [numberOfInputs]);

  const handleInputTextChange = (text, index) => {
    if (!text.length) {
      handleClearInput(index);
    }

    if (text.length > 1) {
      handleClearInput(index);
      Keyboard.dismiss();
      return fillInputs(text);
    }

    if (text) {
      dispatch({
        type: ACTION_TYPES.setOtpTextForIndex,
        payload: {
          text,
          index
        }
      });
      focusInput(index + 1);
    }

    if (index === numberOfInputs - 1 && text) {
      Keyboard.dismiss();
    }
  };

  const _handleTextChange = (text, index) => {
    if (Platform.OS === 'android' && !hasKeySupport || // Pasted from input accessory
    Platform.OS === 'ios' && text.length > 1) {
      handleInputTextChange(text, index);
    }
  };

  const _handleKeyPress = ({
    nativeEvent: {
      key
    }
  }, index) => {
    handleInputTextChange(key === 'Backspace' ? '' : key, index);
    if (Platform.OS === 'android' && !hasKeySupport && !isNaN(parseInt(key))) dispatch({
      type: ACTION_TYPES.setHasKeySupport,
      payload: true
    });
  };

  const focusInput = useCallback(index => {
    if (index >= 0 && index < numberOfInputs) {
      var _input$current2;

      const input = inputs.current[index];
      input === null || input === void 0 ? void 0 : (_input$current2 = input.current) === null || _input$current2 === void 0 ? void 0 : _input$current2.focus();
    }
  }, [numberOfInputs]);
  const handleClearInput = useCallback(inputIndex => {
    var _input$current3;

    const input = inputs.current[inputIndex];
    input === null || input === void 0 ? void 0 : (_input$current3 = input.current) === null || _input$current3 === void 0 ? void 0 : _input$current3.clear();
    dispatch({
      type: ACTION_TYPES.setOtpTextForIndex,
      payload: {
        index: inputIndex,
        text: ''
      }
    });
    focusInput(inputIndex - 1);
  }, [focusInput]);
  const fillInputs = useCallback(code => {
    dispatch({
      type: ACTION_TYPES.setOtpCode,
      payload: {
        numberOfInputs,
        code
      }
    });
  }, [numberOfInputs]);
  const listenOnCopiedText = useCallback(async () => {
    const copiedText = await Clipboard.getString();
    const otpCodeValue = Object.values(otpCode).join('');

    if (copiedText && copiedText.length === numberOfInputs && copiedText !== otpCodeValue && copiedText !== previousCopiedText.current) {
      previousCopiedText.current = copiedText;
      fillInputs(copiedText);
    }
  }, [fillInputs, numberOfInputs, otpCode]);
  useEffect(() => {
    let interval;

    if (autofillFromClipboard) {
      interval = setInterval(() => {
        listenOnCopiedText();
      }, 500);
    }

    return () => {
      clearInterval(interval);
    };
  }, [autofillFromClipboard, listenOnCopiedText, numberOfInputs]);

  const renderInputs = () => {
    const iterationArray = Array(numberOfInputs).fill(0);
    return iterationArray.map((_, index) => {
      let inputIndex = index;

      if (isRTL) {
        inputIndex = numberOfInputs - 1 - index;
      }

      const inputValue = otpCode["".concat(inputIndex)];

      if (!inputs.current[inputIndex]) {
        inputs.current[inputIndex] = /*#__PURE__*/React.createRef();
      }

      return /*#__PURE__*/React.createElement(OtpInput, _extends({
        autoCapitalize: autoCapitalize,
        clearTextOnFocus: clearTextOnFocus,
        firstInput: index === 0,
        focusStyles: focusStyles,
        handleKeyPress: keyPressEvent => _handleKeyPress(keyPressEvent, inputIndex),
        handleTextChange: text => _handleTextChange(text, inputIndex),
        inputContainerStyles: inputContainerStyles,
        inputStyles: inputStyles,
        inputValue: inputValue,
        key: inputIndex,
        keyboardType: keyboardType,
        maxLength: Platform.select({
          android: 1,
          ios: index === 0 ? numberOfInputs : 1
        }),
        numberOfInputs: numberOfInputs,
        placeholder: placeholder // @ts-ignore
        ,
        ref: inputs.current[inputIndex],
        secureTextEntry: secureTextEntry,
        selectTextOnFocus: selectTextOnFocus,
        testID: "".concat(testIDPrefix, "-").concat(inputIndex)
      }, restProps));
    });
  };

  return /*#__PURE__*/React.createElement(View, {
    style: style || styles.container
  }, renderInputs());
});
export { OtpInputsRef };
export default OtpInputs;
